packages/api-client-core/src/operationRunners.ts

```diff
@@ -302,23 +307,54 @@ export const actionRunner: ActionRunner = async <Shape extends RecordShape = any
   }
 };
 
+const processBulkActionResponse = <Shape extends RecordShape = any>(
+  defaultSelection: FieldSelection | null,
+  response: any,
+  records: any,
+  modelSelectionField: string,
+  hasReturnType?: HasReturnType | null
+) => {
+  if (defaultSelection == null) return [];
+  if (!hasReturnType) {
+    return hydrateRecordArray<Shape>(response, records[modelSelectionField]);
+  } else if (typeof hasReturnType == "boolean") {
+    return records.results;
+  } else {
+    return Object.keys(hasReturnType).flatMap((returnTypeField) => {
+      const results = records[returnTypeField];
+
+      return Array.isArray(results) ? results : [];
+    });
```

Comment:
Might need a small refresher here.

I thought the object-y `hasReturnType` would be for the upsert case, where we have something like:

```gql
... on CreateResult { thing { … } }
... on UpdateResult { thing { … } }
```

And the `hasReturnType` would be something like:

```ts
{
  "... on CreateResult": { hasReturnType: false },
  "... on UpdateResult": { hasReturnType: false },
}
```

Is that right? If so, those keys aren't actual fields on the results. I think you'd have to inspect the `__typename` to know how to process the result.

Reply:
updated to inspect the `__typenames`; for the bulk case it's a little more annoying; if there is a return type the query looks like:

```
mutation {
  bulkUpsertPosts(inputs: [{}, {post: {id: "56"}}]) {
  	posts {
      __typename
      ... on UpsertPostReturnType {
        __typename
        result
      }
      ... on Post {
        __typename
        id
      }
    }
  }
}
```

Reply:
🚀 yeah, you have this extra layer to work through which is annoying